drivers/dma/dma_sf32lb.c

```diff
@@ -0,0 +1,425 @@
+			config->n_requests);
+		return -EINVAL;
+	}
+
+	if (config_dma->channel_priority > DMAC_MAX_PL) {
+		LOG_ERR("Invalid channel priority (%" PRIu32 ", max %" PRIu32 ")",
+			config_dma->channel_priority, DMAC_MAX_PL);
+		return -EINVAL;
+	}
+
+	if ((config_dma->head_block->source_addr_adj == DMA_ADDR_ADJ_DECREMENT) |
+	    (config_dma->head_block->dest_addr_adj == DMA_ADDR_ADJ_DECREMENT)) {
+		LOG_ERR("Address decrement not supported");
+		return -ENOTSUP;
+	}
+
+	if ((config_dma->source_data_size != 1U) && (config_dma->source_data_size != 2U) &&
+	    (config_dma->source_data_size != 4U)) {
+		LOG_ERR("Invalid source data size (%" PRIu32 ", must be 1, 2, or 4)",
+			config_dma->source_data_size);
+		return -EINVAL;
+	}
+
+	if ((config_dma->dest_data_size != 1U) && (config_dma->dest_data_size != 2U) &&
+	    (config_dma->dest_data_size != 4U)) {
+		LOG_ERR("Invalid destination data size (%" PRIu32 ", must be 1, 2, or 4)",
+			config_dma->dest_data_size);
+		return -EINVAL;
+	}
+
+	/* configure transfer parameters */
+	ccrx = sys_read32(config->dmac + DMAC_CCRX(channel));
+	ccrx &= ~(DMAC_CCR1_TCIE | DMAC_CCR1_HTIE | DMAC_CCR1_TEIE | DMAC_CCR1_DIR_Msk |
+		  DMAC_CCR1_CIRC_Msk | DMAC_CCR1_PINC_Msk | DMAC_CCR1_MINC_Msk |
+		  DMAC_CCR1_PSIZE_Msk | DMAC_CCR1_MSIZE_Msk | DMAC_CCR1_PL_Msk |
+		  DMAC_CCR1_MEM2MEM_Msk);
+
+	ccrx |= FIELD_PREP(DMAC_CCR1_PL_Msk, config_dma->channel_priority);
+
+	switch (config_dma->channel_direction) {
+	case MEMORY_TO_MEMORY:
+		ccrx |= DMAC_CCR1_MEM2MEM;
+		__fallthrough;
+	case PERIPHERAL_TO_MEMORY:
+		ccrx |= DMAC_CCRX_PSIZE(config_dma->source_data_size) |
+			DMAC_CCRX_MSIZE(config_dma->dest_data_size);
+
+		if (config_dma->head_block->source_addr_adj == DMA_ADDR_ADJ_INCREMENT) {
+			ccrx |= DMAC_CCR1_PINC;
+		}
+		if (config_dma->head_block->dest_addr_adj == DMA_ADDR_ADJ_INCREMENT) {
+			ccrx |= DMAC_CCR1_MINC;
+		}
+
+		cparx = config_dma->head_block->source_address;
+		cm0arx = config_dma->head_block->dest_address;
+		break;
+	case MEMORY_TO_PERIPHERAL:
+		ccrx |= DMAC_CCR1_DIR | DMAC_CCRX_PSIZE(config_dma->dest_data_size) |
+			DMAC_CCRX_MSIZE(config_dma->source_data_size);
+
+		if (config_dma->head_block->source_addr_adj == DMA_ADDR_ADJ_INCREMENT) {
+			ccrx |= DMAC_CCR1_MINC;
+		}
+		if (config_dma->head_block->dest_addr_adj == DMA_ADDR_ADJ_INCREMENT) {
+			ccrx |= DMAC_CCR1_PINC;
+		}
+
+		cparx = config_dma->head_block->dest_address;
+		cm0arx = config_dma->head_block->source_address;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	sys_write32(ccrx, config->dmac + DMAC_CCRX(channel));
+
+	/* single transfer */
+	sys_write32(FIELD_PREP(DMAC_CBSR1_BS_Msk, 0U), config->dmac + DMAC_CBSRX(channel));
+
+	/* configure transfer size, src/dst addresses */
+	sys_write32(config_dma->head_block->block_size, config->dmac + DMAC_CNDTRX(channel));
+	sys_write32(cparx, config->dmac + DMAC_CPARX(channel));
+	sys_write32(cm0arx, config->dmac + DMAC_CM0ARX(channel));
+
+	/* configure request */
+	if (channel < 4U) {
```

Comment:
This is an unfortunate hardware design choice. Because the following section is mutating a register shared among channels from what I can tell, it should be put in a critical section (spin lock).

It's possible, though unlikely, that being preempted and starting a channel from another context could occur here from another interrupt or thread resulting in a race on the cselrx value.

Reply:
It’s less about isr safety and the channels are uniquely owned. If any registers overlap you’ll need to wrap those register updates in locks.

Reply:
yeah, as you said it's unfortunate as everything else does not suffer from this problem...